### PR TITLE
CLDR-17339 Fix problem in ChartDeltaDtds that was skipping keyboard3

### PR DESCRIPTION
**Pull request by macchiati**
CLDR-17339

- [ ] This PR completes the ticket.

ALLOW_MANY_COMMITS=true

### CODE REVIEW

**Comment by macchiati** [comment]
Question for you, @srl295 . There was a keyboard3 in v44. The change here fixes a bug, but also only goes back to v45 (eg no comparison to v44). That's ok if we want to forget about the v44 dtd.

If we want to retain the comparison we would change the following to have 44.
```
    keyboard3("keyboards/dtd/ldmlKeyboard3.dtd", "45.0", null, "../keyboards/3.0"),
    keyboardTest3("keyboards/dtd/ldmlKeyboardTest3.dtd", "45.0", null, "../keyboards/test");
```
Second question is: we are sometimes calling this Keyboard 2.0 (since it is a breaking change), but we have keyboard3. Should we publicly call it Keyboard 3.0?

**Comment by srl295** [comment]
> Question for you, @srl295 . There was a keyboard3 in v44. The change here fixes a bug, but also only goes back to v45 (eg no comparison to v44). That's ok if we want to forget about the v44 dtd.

I don't think there's any value to comparing to v44 here.

> If we want to retain the comparison we would change the following to have 44.
> 
> ```
>     keyboard3("keyboards/dtd/ldmlKeyboard3.dtd", "45.0", null, "../keyboards/3.0"),
>     keyboardTest3("keyboards/dtd/ldmlKeyboardTest3.dtd", "45.0", null, "../keyboards/test");
> ```
> 
> Second question is: we are sometimes calling this Keyboard 2.0 (since it is a breaking change),

I don't think it's being called 2.0 now.

> but we have keyboard3. Should we publicly call it Keyboard 3.0?

It's somewhat of the third revision, with 2.0 being CLDR-10926 (2018).

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/ChartDtdDelta.java [hunk 1]
```diff
@@ -152,7 +152,9 @@ public void writeContents(FormattedFileWriter pw) throws IOException {
                 String firstVersion = type.firstVersion; // FIRST_VERSION.get(type);
                 if (firstVersion != null
                         && current != null
-                        && current.compareTo(firstVersion) < 0) {
+                        && VersionInfo.getInstance(current)
+                                        .compareTo(VersionInfo.getInstance(firstVersion))
+                                < 0) {
```
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/tool/ChartDtdDelta.java:157]: good catch, needed to be a version compare not a lexical compare

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/ChartDtdDelta.java [hunk 2]
```diff
@@ -174,7 +176,11 @@ public void writeContents(FormattedFileWriter pw) throws IOException {
                     continue;
                 }
                 DtdData dtdLast = null;
-                if (last != null && (firstVersion == null || last.compareTo(firstVersion) >= 0)) {
+                if (last != null
+                        && (firstVersion == null
+                                || VersionInfo.getInstance(last)
+                                                .compareTo(VersionInfo.getInstance(firstVersion))
+                                        >= 0)) {
```
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/tool/ChartDtdDelta.java:183]: same

**Review by srl295 — APPROVED** [review]

**Comment by macchiati** [comment]
Ok then this should work

**Pull request merged**